share/p2p/peers/manager.go

```diff
@@ -243,39 +248,54 @@ func (m *Manager) subscribeHeader(ctx context.Context, headerSub libhead.Subscri
 			log.Errorw("get next header from sub", "err", err)
 			continue
 		}
+		m.validatedPool(h.DataHash.String())
 
-		if m.getOrCreatePool(h.DataHash.String()).markValidated() {
-			log.Debugw("marked validated", "datahash", h.DataHash.String())
+		// store first header for validation purposes
+		if atomic.CompareAndSwapUint64(&m.initialHeight, 0, uint64(h.Height())) {
+			log.Debugw("stored initial height", "height", h.Height())
 		}
 	}
 }
 
-// Validate will collect peer.ID into corresponding peer pool
-func (m *Manager) validate(ctx context.Context, peerID peer.ID, hash share.DataHash) pubsub.ValidationResult {
-	if hash.IsEmptyRoot() {
-		// we don't send empty EDS data hashes, but If someone sent it to us - do hard reject
-		return pubsub.ValidationReject
-	}
-
+// validate will collect peer.ID into corresponding peer pool
+func (m *Manager) validate(_ context.Context, peerID peer.ID, msg shrexsub.Notification) pubsub.ValidationResult {
 	// messages broadcast from self should bypass the validation with Accept
 	if peerID == m.host.ID() {
-		log.Debugw("received datahash from self", "datahash", hash.String())
+		log.Debugw("received datahash from self", "datahash", msg.DataHash.String())
 		return pubsub.ValidationAccept
 	}
 
 	// punish peer for sending invalid hash if it has misbehaved in the past
-	if m.hashIsBlacklisted(hash) {
-		log.Debugw("received blacklisted hash, reject validation", "peer", peerID, "datahash", hash.String())
+	if m.hashIsBlacklisted(msg.DataHash) {
+		log.Debugw("received blacklisted hash, reject validation", "peer", peerID, "datahash", msg.DataHash.String())
 		return pubsub.ValidationReject
 	}
 
-	if m.peerIsBlacklisted(peerID) {
-		log.Debugw("received message from blacklisted peer, reject validation", "peer", peerID, "datahash", hash.String())
+	if m.isBlacklistedPeer(peerID) {
+		log.Debugw("received message from blacklisted peer, reject validation",
+			"peer", peerID,
+			"datahash", msg.DataHash.String())
 		return pubsub.ValidationReject
 	}
 
-	m.getOrCreatePool(hash.String()).add(peerID)
-	log.Debugw("got hash from shrex-sub", "peer", peerID, "datahash", hash.String())
+	if msg.Height == 0 {
+		log.Debugw("received message with 0 height", "peer", peerID)
+		return pubsub.ValidationReject
+	}
+
+	if msg.Height < atomic.LoadUint64(&m.initialHeight) {
+		// we can use peers from discovery for headers before the first one from headerSub
+		// if we allow pool creation for those headers, there is chance the pool will not be validated in
+		// time and will be false-positively trigger blacklisting of hash and all peers that sent msgs for
+		// that hash
+		log.Debugw("received message for past header", "peer", peerID, "datahash", msg.DataHash.String())
+		return pubsub.ValidationIgnore
```

Comment:
We should do hard reject for most heights unless it's recent. At this point, it does not matter as for FloodSub the Ignore and the Reject are effectively the same, but it's likely that we will move to a different router at some point. It's also just nice to codify the exact rules, what is hard reject, and what is acceptable via ignore

Reply:
In case of messages handled by this rule, we we don't target any peer scoring or filtering mechanics, just want to address corner case of subjective start and doing as little logic as possible. The reason why `pubsub.ValidationIgnore` is used here is because there is no strong evidence of any misbehave from peer and thus no punishment needed.